core/client/fs/src/main/java/alluxio/client/block/stream/DefaultBlockWorkerClient.java

```diff
@@ -38,207 +40,241 @@
-        close();
-        userState.relogin();
-        lastException = e;
-      }
-    }
-    if (lastException != null) {
-      throw lastException;
-    }
-    mStreamingAsyncStub = BlockWorkerGrpc.newStub(mStreamingChannel);
-    mRpcBlockingStub = BlockWorkerGrpc.newBlockingStub(mRpcChannel);
-    mRpcFutureStub = BlockWorkerGrpc.newFutureStub(mRpcChannel);
+  public DefaultBlockWorkerClient(ClientContext context, GrpcServerAddress address,
+                                  AlluxioConfiguration alluxioConf) {
+    // BlockWorkerClient is typically used inside AlluxioFileInStream
+    // to fetch block chunk. AlluxioFilInStream has its own retry mechanism, so we don't
+    // retry at this level so as not to interfere with the retry of our higher-level
+    // callers.
+    // An instance where retries interfere with each other is when we and our calling
+    // AlluxioFileInStream both adopted a timer-based retry, then all the time could be wasted
+    // in our inner retry, and the outer stream has no chance of retrying a different client.
+    super(context, () -> new CountingRetry(0));
+    // the server address this client connects to. It might not be an IP address
     mAddress = address;
     mRpcTimeoutMs = alluxioConf.getMs(PropertyKey.USER_RPC_RETRY_MAX_DURATION);
     mTracker = DETECTOR.track(this);
   }
 
   @Override
-  public boolean isShutdown() {
-    return mStreamingChannel.isShutdown() || mRpcChannel.isShutdown();
+  protected void beforeConnect() {}
+
+  @Override
+  protected void afterConnect() {
+    // set up stubs
+    mStreamingAsyncStub = BlockWorkerGrpc.newStub(mStreamingChannel);
+    mRpcFutureStub = BlockWorkerGrpc.newFutureStub(mChannel);
+    mRpcBlockingStub = BlockWorkerGrpc.newBlockingStub(mChannel);
   }
 
   @Override
-  public boolean isHealthy() {
-    return !isShutdown() && mStreamingChannel.isHealthy() && mRpcChannel.isHealthy();
+  protected void doConnect() throws AlluxioStatusException {
+    super.doConnect();
+
+    // build a streaming channel
+    AlluxioConfiguration conf = mContext.getClusterConf();
+    mStreamingChannel = GrpcChannelBuilder.newBuilder(mServerAddress, conf)
+            .setSubject(mContext.getSubject())
+            .setNetworkGroup(GrpcNetworkGroup.STREAMING)
+            .build();
   }
 
   @Override
-  public void close() throws IOException {
-    try (Closer closer = Closer.create()) {
-      closer.register(() -> {
-        if (mStreamingChannel != null) {
-          mStreamingChannel.shutdown();
-        }
-      });
-      closer.register(() -> {
-        if (mRpcChannel != null) {
-          mRpcChannel.shutdown();
-        }
-      });
-      closer.register(() -> {
-        if (mTracker != null) {
-          mTracker.close(this);
-        }
-      });
+  protected void shutdownChannelsIfCreated() {
+    super.shutdownChannelsIfCreated();
+    if (mStreamingChannel != null) {
+      mStreamingChannel.shutdown();
+      mStreamingChannel = null;
     }
```

Comment:
This `null` check is necessary because this method will be called on partial failures, so we can't be sure which channels are established and which are not.